Задания по 5-му спринту

Danzl opened the pull request:
Логика работы приложения разбита на несколько частей. Работа по генерации вопросов для квиза вынесена в отдельную сущность. Часть логики приложения вынесена в отдельный класс ResultAlertPresenter. Обработка нажатия на кнопку алерт реализована.
Реализована функция сохранения лучшего результата store (с проверкой на то, что новый лучший результат сохранён в User Defaults). К самому алерту в поле message передан сформированный из рекордной игры текст. Приложение выглядит корректно — элементы интерфейса адаптируются под разрешения экранов iPhone, начиная с X. По окончании раунда вызывается алерт, который соответствует макету и на котором отображаются следующие данные: Текущий результат в виде соотношения правильно отвеченных в раунде вопросов и общего количества вопросов в одном раунде Количество сыгранных квизов за всё время, что приложение установлено Лучший результат за всё время, что приложение установлено. В той же строке должна выводится дата достижения этого результата в формате согласно макетам Средняя точность, рассчитывающаяся как отношение правильно отвеченных вопросов за всё время игры к общему количеству вопросов за всё время игры. Точность должна быть выражена в процентах. Класс StatisticServiceImplementation закрыт протоколом StatisticService. В MovieQuizViewController используется переменная statisticService типа StatisticService. В коде нет ретейн циклов.
При завершении приложения результаты, которые показываются в конце раунда, не обнуляются.

Danzl committed:
Задания по 5-му спринту


Логика работы приложения разбита на несколько частей.
Работа по генерации вопросов для квиза вынесена в отдельную сущность.
Часть логики приложения вынесена в отдельный класс ResultAlertPresenter.
Обработка нажатия на кнопку алерт реализована.
Реализована функция сохранения лучшего результата store (с проверкой на то, что новый лучший результат сохранён в User Defaults). К самому алерту в поле message передан сформированный из рекордной игры текст.
Приложение выглядит корректно — элементы интерфейса адаптируются под разрешения экранов iPhone, начиная с X.
По окончании раунда вызывается алерт, который соответствует макету и на котором отображаются следующие данные:
Текущий результат в виде соотношения правильно отвеченных в раунде вопросов и общего количества вопросов в одном раунде
Количество сыгранных квизов за всё время, что приложение установлено
Лучший результат за всё время, что приложение установлено. В той же строке должна выводится дата достижения этого результата в формате согласно макетам
Средняя точность, рассчитывающаяся как отношение правильно отвеченных вопросов за всё время игры к общему количеству вопросов за всё время игры. Точность должна быть выражена в процентах.
Класс StatisticServiceImplementation закрыт протоколом StatisticService. В MovieQuizViewController используется переменная statisticService типа StatisticService.
В коде нет ретейн циклов.
При завершении приложения результаты, которые показываются в конце раунда, не обнуляются.

ABashkirova commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -0,0 +1,32 @@
+    func showResultsAlert() {
+        let alert = UIAlertController(title: alertModel.title, message: alertModel.message, preferredStyle: .alert)
+        let action = UIAlertAction(title: alertModel.buttonText, style: .default) { _ in
+            self.alertModel.completion()
```
здесь может произойти утечка памяти, лучше захватить слабо ссылку. 
```suggestion
        let action = UIAlertAction(title: alertModel.buttonText, style: .default) { [weak self] _ in
            self?.alertModel.completion()
```

ABashkirova commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,38 @@
+
+class QuestionFactory: QuestionFactoryProtocol {
+    weak var delegate: QuestionFactoryDelegate?
+    init(delegate: QuestionFactoryDelegate?) {
```
```suggestion

    init(delegate: QuestionFactoryDelegate?) {
```

ABashkirova commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,38 @@
+    weak var delegate: QuestionFactoryDelegate?
+    init(delegate: QuestionFactoryDelegate?) {
+        self.delegate = delegate
+    }
```
```suggestion
    }
    
```

ABashkirova commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,81 @@
+    
+    var gamesCount: Int {
+        get { userDefaults.integer(forKey: Keys.gamesCount.rawValue) }
+        set { userDefaults.set(newValue,  forKey: Keys.gamesCount.rawValue) }
```
```suggestion
        set { userDefaults.set(newValue, forKey: Keys.gamesCount.rawValue) }
```

ABashkirova commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -0,0 +1,198 @@
+        showAnswerResult(isCorrect: false == currentQuestion.correctAnswer)
+    }
+    
+    func didRecieveNextQuestion(question: QuizQuestion?) {
```
Рекомендую включить проверку правописания, чтобы избежать опечаток :)
https://sarunw.com/posts/spell-checking-in-xcode/

ABashkirova commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -0,0 +1,198 @@
+ Настоящий рейтинг: 9,2
+ Вопрос: Рейтинг этого фильма больше чем 6?
+ Ответ: ДА
+
```
Наверное, уже можно удалять

ABashkirova commented on MovieQuiz-ios-project_sprint_3_start-3/MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,81 @@
+    
+    var totalAccuracy : Double {
+        get {
+            return Double(totalScore) / Double(gamesCount) * 10.0
```
10.0 - магическое число :) обычно такое задают константой, чтобы было понятно из какого технического требования оно появилось

Плюс, если эта константа общая для всего приложения, не для одного сервиса, то хорошо бы ее хранить в одном месте, а все остальные сервисы от нее настраивать.

ABashkirova reviewed (approved):
Привет!

Нужно исправить замечания под код стайлу, небольшую ошибку с захватом ссылки.

Danzl committed:
Исправления с ошибкой захвата и по код стайлу.

Исправления по код стайлу: удалены лишние строки, добавлены отступы где нужно.
Захвачена слабо ссылка в AlertPresenter.
Спасибо за совет по правописанию =)
